libpod/image/utils.go

```diff
@@ -29,23 +33,50 @@ func findImageInRepotags(search imageParts, images []*Image) (*storage.Image, er
 				continue
 			}
 			_, dName, dSuspiciousTagValueForSearch := d.suspiciousRefNameTagValuesForSearch()
-			if dName == searchName && dSuspiciousTagValueForSearch == searchSuspiciousTagValueForSearch {
-				results = append(results, image.image)
+			if dSuspiciousTagValueForSearch != searchSuspiciousTagValueForSearch {
 				continue
 			}
-			// account for registry:/somedir/image
-			if strings.HasSuffix(dName, "/"+searchName) && dSuspiciousTagValueForSearch == searchSuspiciousTagValueForSearch {
-				results = append(results, image.image)
-				continue
+			c := Candidate{
+				name:  name,
+				image: image,
+			}
+			if dName == searchName || strings.HasSuffix(dName, "/"+searchName) {
+				candidates = append(candidates, c)
 			}
 		}
 	}
-	if len(results) == 0 {
-		return &storage.Image{}, errors.Errorf("unable to find a name and tag match for %s in repotags", searchName)
-	} else if len(results) > 1 {
-		return &storage.Image{}, errors.Wrapf(define.ErrMultipleImages, searchName)
+	if len(candidates) == 1 {
+		return candidates[0].image.image, nil
+	}
+	// If candidates all have same name and only one is read/write return it.
+	// Othewise return error with the list of candidates
+	if len(candidates) > 1 {
+		var (
+			rwName  string
+			rwImage *Image
+		)
+		names := make(map[string]bool)
+		for _, c := range candidates {
+			names[c.name] = true
+			if !c.image.IsReadOnly() {
+				if rwImage != nil {
+					return nil, errors.Wrapf(define.ErrMultipleImages, "multiple read/write images found %s and %s", rwName, c.name)
+				}
+				rwImage = c.image
+				rwName = c.name
+			}
+		}
+		// If only one name used and have read/write image return it
+		if len(names) == 1 && rwImage != nil {
+			return rwImage.image, nil
+		}
+		keys := []string{}
+		for k := range names {
+			keys = append(keys, k)
+		}
+		return nil, errors.Wrapf(define.ErrMultipleImages, "found %s", keys)
 	}
-	return results[0], nil
+	return nil, errors.Errorf("unable to find a name and tag match for %s in repotags", searchName)
```

Comment:
(Absolutely non-blocking: I found the previous version, with the `len() == 0` case at the front, a bit more readable. All the conditions are together; the way it is now, I have to scroll back and see what this case is about.

Reply:
Rearranged.